tt_eager/tt_dnn/op_library/untilize/kernels/dataflow/writer_unary_unpad_batch_rows_sharded.cpp

```diff
@@ -21,8 +23,13 @@ void kernel_main() {
     for(uint32_t b = 0; b < batch; ++b) {
         cb_wait_front(cb_id_untilize_out, num_padded_tiles_per_batch);
         uint64_t noc_l1_read_addr = get_noc_addr(get_read_ptr(cb_id_untilize_out));
-        noc_async_read(noc_l1_read_addr, l1_write_addr, unpadded_block_size_bytes);
-        l1_write_addr += unpadded_block_size_bytes;
+
+        for (uint32_t row = 0; row < num_padded_rows_per_batch; ++row) {
+            noc_async_read(noc_l1_read_addr, l1_write_addr, unpadded_block_row_size_bytes);
+            noc_l1_read_addr += padded_block_row_size_bytes;
+            l1_write_addr += unpadded_block_row_size_bytes;
+        }
+
         noc_async_read_barrier();
```

Comment:
Can we move this barrier outside of the batch loop?

Reply:
We can't move it outside, because we need it to protect the state of `noc_l1_read_addr`, which will be updated on the next batch.

Reply:
If we fully buffer `cb_id_untilize_out` then it could be moved outside. Maybe if you see this is more performant and it fits for your use case you could expose an option or something for this?

Something you could also try to see if perf is better is to switch from using read to using write, and then using noc_async_writes_flushed() as the barrier instead (I don't know if it'll be better or worse)

Reply:
These are great suggestions, but I'd prefer to do optimizations in a separate PR, if you're OK with it.

Reply:
Agree! Let's do it as follow on